src/search/landmarks/exploration.cc

```diff
@@ -147,124 +122,91 @@ void Exploration::build_unary_operators(const OperatorProxy &op) {
             || excluded_op_ids.count(op.op_or_axiom_id)) {
-            op.h_add_cost = -2; // operator will not be applied during relaxed exploration
+            // operator will not be applied during relaxed exploration
+            op.excluded = true;
             continue;
         }
-        op.h_add_cost = op.base_cost; // will be increased by precondition costs
         op.h_max_cost = op.base_cost;
-        op.depth = -1;
 
         if (op.unsatisfied_preconditions == 0) {
-            op.depth = 0;
-            int depth = op.is_induced_by_axiom(task_proxy) ? 0 : 1;
-            enqueue_if_necessary(op.effect, op.base_cost, depth, &op, use_h_max);
+            enqueue_if_necessary(op.effect, op.base_cost, &op);
         }
     }
 }
 
-void Exploration::relaxed_exploration(bool use_h_max, bool level_out) {
-    int unsolved_goals = termination_propositions.size();
+void Exploration::relaxed_exploration() {
     while (!prop_queue.empty()) {
         pair<int, ExProposition *> top_pair = prop_queue.pop();
         int distance = top_pair.first;
         ExProposition *prop = top_pair.second;
 
-        int prop_cost;
-        if (use_h_max)
-            prop_cost = prop->h_max_cost;
-        else
-            prop_cost = prop->h_add_cost;
+        int prop_cost = prop->h_max_cost;
         assert(prop_cost <= distance);
         if (prop_cost < distance)
             continue;
-        if (!level_out && prop->is_termination_condition && --unsolved_goals == 0)
-            return;
         const vector<ExUnaryOperator *> &triggered_operators = prop->precondition_of;
         for (size_t i = 0; i < triggered_operators.size(); ++i) {
             ExUnaryOperator *unary_op = triggered_operators[i];
-            if (unary_op->h_add_cost == -2) // operator is not applied
+            if (unary_op->excluded) // operator is not applied
                 continue;
             --unary_op->unsatisfied_preconditions;
-            increase_cost(unary_op->h_add_cost, prop_cost);
             unary_op->h_max_cost = max(prop_cost + unary_op->base_cost,
                                        unary_op->h_max_cost);
-            unary_op->depth = max(unary_op->depth, prop->depth);
             assert(unary_op->unsatisfied_preconditions >= 0);
             if (unary_op->unsatisfied_preconditions == 0) {
-                int depth = unary_op->is_induced_by_axiom(task_proxy)
-                    ? unary_op->depth : unary_op->depth + 1;
-                if (use_h_max)
-                    enqueue_if_necessary(unary_op->effect, unary_op->h_max_cost,
-                                         depth, unary_op, use_h_max);
-                else
-                    enqueue_if_necessary(unary_op->effect, unary_op->h_add_cost,
-                                         depth, unary_op, use_h_max);
+                enqueue_if_necessary(unary_op->effect, unary_op->h_max_cost,
+                                     unary_op);
             }
         }
     }
```

Comment:
We already have an implementation of (qualitative) relaxed reachability in the LM-Cut code, I think. In revising this, it might make sense to look into that because it's more optimized. IIRC, it is a depth-first search implemented via recursion rather than an open list.